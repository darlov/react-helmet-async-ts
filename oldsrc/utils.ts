import {TAG_NAMES, TAG_PROPERTIES, ATTRIBUTE_NAMES, SEO_PRIORITY_TAGS} from './constants';
import {
  HelmetProps,
  BaseProps,
  AttributeNamesMap,
  TagPropsType,
  IStateType,
  MetaProps,
  LinkProps,
  ScriptProps
} from "./types";

enum HELMET_PROPS {
  DEFAULT_TITLE = 'defaultTitle',
  DEFER = 'defer',
  ENCODE_SPECIAL_CHARACTERS = 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE = 'onChangeClientState',
  TITLE_TEMPLATE = 'titleTemplate',
  PRIORITIZE_SEO_TAGS = 'prioritizeSeoTags'
}

const getInnermostProperty = <T extends keyof HelmetProps>(propsList: HelmetProps[], property: T) => {
  for (let i = propsList.length - 1; i >= 0; i -= 1) {
    const props = propsList[i];

    if (Object.hasOwn(props, property)) {
      return props[property];
    }
  }

  return undefined;
};

const getTitleFromPropsList = (propsList: HelmetProps[]) => {
  let innermostTitle = getInnermostProperty(propsList, TAG_NAMES.TITLE);
  const innermostTemplate = getInnermostProperty(propsList, HELMET_PROPS.TITLE_TEMPLATE);
  if (Array.isArray(innermostTitle)) {
    innermostTitle = innermostTitle.join('');
  }
  if (innermostTemplate && innermostTitle) {
    // use function arg to avoid need to escape $ characters
    return innermostTemplate.replace(/%s/g, innermostTitle);
  }

  const innermostDefaultTitle = getInnermostProperty(propsList, HELMET_PROPS.DEFAULT_TITLE);

  return innermostTitle || innermostDefaultTitle || undefined;
};

const getOnChangeClientState = (propsList: HelmetProps[]) =>
  getInnermostProperty(propsList, HELMET_PROPS.ON_CHANGE_CLIENT_STATE) || (() => {
  });

const getAttributesFromPropsList = <T extends ATTRIBUTE_NAMES>(tagType: T, propsList: HelmetProps[]) =>
  propsList
    .map(props => props[tagType])
    .filter((tag): tag is AttributeNamesMap[T] => !!tag)
    .reduce<AttributeNamesMap[T]>((tagAttrs, current) => ({...tagAttrs, ...current}), {});

const getBaseTagFromPropsList = (primaryAttributes: TAG_PROPERTIES[], propsList: HelmetProps[]) =>
  propsList
    .map(props => props.base)
    .filter((tag): tag is BaseProps => !!tag)
    .reverse()
    .reduce<BaseProps[]>((innermostBaseTag, tag) => {
      if (!innermostBaseTag.length) {
        const keys = Object.keys(tag);

        for (let i = 0; i < keys.length; i += 1) {
          const attributeKey = keys[i];
          const lowerCaseAttributeKey = attributeKey.toLowerCase();

          if (primaryAttributes.some(m => m === lowerCaseAttributeKey) && tag[lowerCaseAttributeKey as keyof BaseProps]) {
            return innermostBaseTag.concat(tag);
          }
        }
      }

      return innermostBaseTag;
    }, []);

// eslint-disable-next-line no-console
const warn = (msg: string) => console && typeof console.warn === 'function' && console.warn(msg);

type Flatten<T> = T extends any[] ? T[number] : T;

const getTagsFromPropsList = <T extends keyof TagPropsType,
  TValidTags extends Required<TagPropsType>,
  TTags = TagPropsType[T],
  TElement = Flatten<TTags>>(tagName: T, primaryAttributes: TAG_PROPERTIES[], propsList: TagPropsType[]) => {
  // Calculate list of tags, giving priority innermost component (end of the propslist)
  const approvedSeenTags: { [key in keyof TElement]: any; } = {} as TElement;

  return propsList
    .filter((props): props is TValidTags => {
      if (Array.isArray(props[tagName])) {
        return true;
      }
      if (typeof props[tagName] !== 'undefined') {
        warn(
          `Helmet: ${ tagName } should be of type "Array". Instead found type "${ typeof props[
            tagName
            ] }"`
        );
      }
      return false;
    })
    .map((props) => props[tagName])
    .reverse()
    .reduce((approvedTags, instanceTags) => {
      const instanceSeenTags: { [key in keyof TElement]: any; } = {} as TElement;
      
      instanceTags
        .filter(tag => {
          let primaryAttributeKey: string | undefined = undefined;
          const keys = Object.keys(tag);
          for (let i = 0; i < keys.length; i += 1) {
            const attributeKey = keys[i];
            const lowerCaseAttributeKey = attributeKey.toLowerCase();

            // Special rule with link tags, since rel and href are both primary tags, rel takes priority
            if (
              primaryAttributes.some(m => m === lowerCaseAttributeKey) &&
              !(
                primaryAttributeKey === TAG_PROPERTIES.REL &&
                tag[primaryAttributeKey as keyof typeof tag].toLowerCase() === 'canonical'
              ) &&
              !(
                lowerCaseAttributeKey === TAG_PROPERTIES.REL &&
                tag[lowerCaseAttributeKey as keyof typeof tag].toLowerCase() === 'stylesheet'
              )
            ) {
              primaryAttributeKey = lowerCaseAttributeKey;
            }
            // Special case for innerHTML which doesn't work lowercased
            if (
              primaryAttributes.some(m => m === attributeKey) &&
              (attributeKey === TAG_PROPERTIES.INNER_HTML ||
                attributeKey === TAG_PROPERTIES.CSS_TEXT ||
                attributeKey === TAG_PROPERTIES.ITEM_PROP)
            ) {
              primaryAttributeKey = attributeKey;
            }
          }

          if (!primaryAttributeKey || !tag[primaryAttributeKey as keyof typeof tag]) {
            return false;
          }

          const value: string = tag[primaryAttributeKey as keyof typeof tag].toLowerCase();

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement]) {
            approvedSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!instanceSeenTags[primaryAttributeKey as keyof TElement]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement][value]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement][value] = true;
            return true;
          }

          return false;
        }, [])
        .reverse()
        .forEach(tag => approvedTags.push(tag as TElement));

      // Update seen tags with tags from this instance
      const keys = Object.keys(instanceSeenTags);
      for (let i = 0; i < keys.length; i += 1) {
        const attributeKey = keys[i];
        const tagUnion = {
          ...approvedSeenTags[attributeKey as keyof TElement],
          ...instanceSeenTags[attributeKey as keyof TElement],
        };

        approvedSeenTags[attributeKey as keyof TElement] = tagUnion;
      }

      return approvedTags;
    }, [] as TElement[])
    .reverse();
};

const getAnyTrueFromPropsList = <T extends keyof HelmetProps>(propsList: HelmetProps[], checkedTag: T) => {
  if (Array.isArray(propsList) && propsList.length) {
    for (let index = 0; index < propsList.length; index += 1) {
      const prop = propsList[index];
      if (prop[checkedTag]) {
        return true;
      }
    }
  }
  return false;
};
const reducePropsToState = (propsList: HelmetProps[]): IStateType => ({
  baseTag: getBaseTagFromPropsList([TAG_PROPERTIES.HREF], propsList),
  bodyAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.BODY, propsList),
  defer: getInnermostProperty(propsList, HELMET_PROPS.DEFER),
  encode: getInnermostProperty(propsList, HELMET_PROPS.ENCODE_SPECIAL_CHARACTERS),
  htmlAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.HTML, propsList),
  linkTags: getTagsFromPropsList(
    TAG_NAMES.LINK,
    [TAG_PROPERTIES.REL, TAG_PROPERTIES.HREF],
    propsList
  ),
  metaTags: getTagsFromPropsList(
    TAG_NAMES.META,
    [
      TAG_PROPERTIES.NAME,
      TAG_PROPERTIES.CHARSET,
      TAG_PROPERTIES.HTTPEQUIV,
      TAG_PROPERTIES.PROPERTY,
      TAG_PROPERTIES.ITEM_PROP,
    ],
    propsList
  ),
  noscriptTags: getTagsFromPropsList(TAG_NAMES.NOSCRIPT, [TAG_PROPERTIES.INNER_HTML], propsList),
  onChangeClientState: getOnChangeClientState(propsList),
  scriptTags: getTagsFromPropsList(
    TAG_NAMES.SCRIPT,
    [TAG_PROPERTIES.SRC, TAG_PROPERTIES.INNER_HTML],
    propsList
  ),
  styleTags: getTagsFromPropsList(TAG_NAMES.STYLE, [TAG_PROPERTIES.CSS_TEXT], propsList),
  title: getTitleFromPropsList(propsList),
  titleAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.TITLE, propsList),
  prioritizeSeoTags: getAnyTrueFromPropsList(propsList, HELMET_PROPS.PRIORITIZE_SEO_TAGS),
});

export const flattenArray = (possibleArray: any) =>
  Array.isArray(possibleArray) ? possibleArray.join('') : possibleArray;

export { reducePropsToState };

const checkIfPropsMatch = (props: any, toMatch: any) => {
  const keys = Object.keys(props);
  for (let i = 0; i < keys.length; i += 1) {
    // e.g. if rel exists in the list of allowed props [amphtml, alternate, etc]
    if (toMatch[keys[i]] && toMatch[keys[i]].includes(props[keys[i]])) {
      return true;
    }
  }
  return false;
};

export type Prioritizer =
  { element: MetaProps, propsToMatch: typeof SEO_PRIORITY_TAGS["meta"] }
  | { element: LinkProps, propsToMatch: typeof SEO_PRIORITY_TAGS["link"] }
  | { element: ScriptProps, propsToMatch: typeof SEO_PRIORITY_TAGS["script"] };

export const prioritizer = <TElement extends Prioritizer["element"], TMatch extends Prioritizer["propsToMatch"]>(
  elements: TElement[],
  propsToMatch: TMatch) => {

  const result: { priority: TElement[], default: TElement[] } = {priority: [], default: []};

  if (Array.isArray(elements)) {

    for (const element of elements) {
      if (checkIfPropsMatch(element, propsToMatch)) {
        result.priority.push(element);
      } else {
        result.default.push(element);
      }
    }
  }
  return  result;
};

export const without = (obj: any, key: any) => {
  return {
    ...obj,
    [key]: undefined,
  };
}
