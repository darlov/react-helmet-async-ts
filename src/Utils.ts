import { Dispatch, SetStateAction } from "react";
import { IHelmetInstanceState, IHelmetTags, TagProps } from "./Types";

export const addUniqueItem = <T, K extends T[] | undefined>(items: K, item: T, keySelector?: (item: T) => any): K => {
  if (!items) {
    return [item] as K;
  }

  const exist = keySelector !== undefined 
      ? items.some(m => keySelector(m) === keySelector(item))
      : items.some(m => m === item);
  return exist ? items : [...items, item] as K;
}

export const removeAction = <T, K extends T[] | undefined>(action: Dispatch<SetStateAction<K>>, keySelector?: (item: T) => any) => (item: T) => {
  action(items => {
    return keySelector !== undefined 
        ? items?.filter(m => keySelector(m) !== keySelector(item)) as K
        : items?.filter(m => m !== item) as K
  });
};
export const addAction = <T, K extends T[] | undefined>(action: Dispatch<SetStateAction<K>>, keySelector?: (item: T) => any) => {
  return (item: T) => action(items => addUniqueItem<T, K>(items, item, keySelector));
};


export namespace _ {
  export type TSelector<T> = ((elem: T) => any) | keyof T;

  export const sortBy = <T>(array: T[] | undefined, selector: TSelector<T>) => {
    if (typeof selector === "function") {
      return array?.sort((x, y) => selector(x) > selector(y) ? 1 : selector(x) < selector(y) ? -1 : 0);
    }

    return array?.sort((x, y) => x[selector] > y[selector] ? 1 : x[selector] < y[selector] ? -1 : 0);
  }

  export const clear = <T>(array: T[] | undefined) => {
    if (array !== undefined && array.length > 0) {
      array.splice(0, array.length)
    }
  }

  export const isEmpty = (obj: Record<string, any>) => {
    return Object.keys(obj).length === 0;
  }

  type TArray = any[] | undefined
  export const isEmptyArray = (...obj: TArray[]) => {
    return obj.every(m => m === undefined || m.length === 0)
  }

  type InferKey<T> = T extends Partial<Record<infer K, any>> ? K : never;
  type InferValue<T> = T extends Partial<Record<any, infer V>> ? V : never;

  export const toPairs = <T extends Partial<Record<string, any>>>(obj: T) => {
    return Object.entries(obj) as [InferKey<T>, InferValue<T>][];
  };


  type TValSelReturn<TItem, T extends ((arg: TItem, index: number) => any) | undefined>
    = T extends (arg: TItem, index: number) => infer K ? K
    : T extends undefined ? TItem
      : never;

  type TValueSelector<T> = ((item: T, index: number) => T) | ((item: T, index: number) => any)

  export const groupBy = <T, K extends keyof any, V extends TValSelReturn<T, TFunc>, TFunc extends TValueSelector<T>>(items: T[], keySelector: (item: T, index: number) => K | undefined, valueSelector?: TFunc): Map<K, TValSelReturn<T, TFunc>[]> => {
    return items.reduce((result, item, index) => {
      const key = keySelector(item, index);
      if (key !== undefined) {
        if (!result.has(key)) {
          result.set(key, [])
        }
        result.get(key)!.push(valueSelector && valueSelector(item, index));
        return result
      }
      return result;
    }, new Map<K, TValSelReturn<T, TFunc>[]>);
  }
}

export const mergeTags = <T extends TagProps>(tagProp: keyof IHelmetTags, result: T[], instances: IHelmetInstanceState[], primaryAttributeSelector: (tag: T) => string | undefined) => {
  for (const instance of instances) {
    const inTags = instance[tagProp] as T[] | undefined;
    if (inTags) {
      if (instance.emptyState) {
        _.clear(result);
      } else if (result.length === 0) {
        result.push(...inTags);
      } else {
        const instanceGrouped = _.groupBy(inTags, primaryAttributeSelector, (item, index) => ([item, index] as [T, number]));
        const resultGrouped = _.groupBy(result, primaryAttributeSelector, (item, index) => ([item, index] as [T, number]));

        for (const [attr, instanceTags] of instanceGrouped.entries()) {
          const resultTags = resultGrouped.get(attr);

          if (resultTags === undefined) {
            result.push(...instanceTags.map(([tags]) => tags))
          } else if (instanceTags.length !== resultTags.length) {
            result = result.filter(m => !resultTags.some(([tag]) => m === tag));
            result.push(...instanceTags.map(([tag]) => tag))
          } else {
            for (let i = 0; i < instanceTags.length; i++) {
              const [instanceTag] = instanceTags[i];
              const [resultTag, resultIndex] = resultTags[i];

              result[resultIndex] = {...resultTag, ...instanceTag};
            }
          }
        }
      }
    }
  }
  
  return result;
}

