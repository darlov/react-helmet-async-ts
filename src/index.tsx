export { HelmetContextProvider, useHelmetContext } from "./HelmetProvider"
export { Helmet } from "./Helmet"
export * from "./tags"
export { HELMET_ATTRIBUTE } from "./Types"