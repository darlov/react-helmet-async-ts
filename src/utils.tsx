import {createElement, Key, ReactElement} from "react";
import {
  ArrayElement,
  DefaultTagPriorityConfig,
  HelmetTags,
  IHelmetInstanceState,
  IHelmetServerState,
  IHelmetState,
  ITypedTagProps,
  primaryLinkAttributes,
  primaryMetaAttributes,
  TagName,
  TagNames,
  TagPriorityConfig,
  TypedTagProps,
} from "./types";
import {renderToStaticMarkup} from "react-dom/server";

export const addUniqueItem = <T, K extends T[] | undefined>(
  items: K,
  item: T,
  keySelector?: (item: T) => any
): K => {
  if (!items) {
    return [item] as K;
  }

  const exist =
    keySelector !== undefined
      ? items.some(m => keySelector(m) === keySelector(item))
      : items.some(m => m === item);
  return exist ? items : ([...items, item] as K);
};

export const removeItem = <T, K extends T[] | undefined>(
  items: K,
  item: T,
  keySelector?: (item: T) => any
): K => {
  return keySelector !== undefined
    ? (items?.filter(m => keySelector(m) !== keySelector(item)) as K)
    : (items?.filter(m => m !== item) as K);
}

export namespace _ {
  export type TSelector<T> = ((elem: T) => any) | keyof T;

  export const sortBy = <T, K extends T[] | undefined>(array: K, selector: TSelector<T>): K => {
    if (array === undefined) {
      return array;
    }

    if (typeof selector === "function") {
      return array.sort((x, y) =>
        selector(x) > selector(y) ? 1 : selector(x) < selector(y) ? -1 : 0
      ) as K;
    }

    return array.sort((x, y) =>
      x[selector] > y[selector] ? 1 : x[selector] < y[selector] ? -1 : 0
    ) as K;
  };

  export const clear = <T, >(array: T[] | undefined) => {
    if (array !== undefined && array.length > 0) {
      array.splice(0, array.length);
    }
  };

  export const isEmpty = (obj: Record<string, any>) => {
    return Object.keys(obj).length === 0;
  };

  type TArray = any[] | undefined;
  export const isEmptyArray = (...obj: TArray[]) => {
    return obj.every(m => m === undefined || m.length === 0);
  };


  type TValSelReturn<
    TItem,
    T extends ((arg: TItem, index: number) => any) | undefined
  > = T extends (arg: TItem, index: number) => infer K ? K : T extends undefined ? TItem : never;

  type TValueSelector<T> = ((item: T, index: number) => T) | ((item: T, index: number) => any);

  export const groupBy = <
    T,
    K extends keyof any,
    V extends TValSelReturn<T, TFunc>,
    TFunc extends TValueSelector<T>
  >(
    items: T[],
    keySelector: (item: T, index: number) => K | undefined,
    valueSelector?: TFunc
  ): Map<K, TValSelReturn<T, TFunc>[]> => {
    return items.reduce((result, item, index) => {
      const key = keySelector(item, index);
      if (key !== undefined) {
        if (!result.has(key)) {
          result.set(key, []);
        }
        result.get(key)!.push(valueSelector && valueSelector(item, index));
        return result;
      }
      return result;
    }, new Map<K, TValSelReturn<T, TFunc>[]>());
  };

  export const remove = <T, >(elements: T[], element: T) => {
    const index = elements.indexOf(element, 0);
    if (index > -1) {
      return elements.splice(index, 1);
    }
    return elements;
  };
}

const findMetaPrimaryAttribute = (meta: ITypedTagProps<TagName.meta>) => {
  const foundAttr = primaryMetaAttributes.find(attr => meta.tagProps[attr] !== undefined);

  if (foundAttr !== undefined) {
    return `${foundAttr}_${meta.tagProps[foundAttr]}`;
  }

  return undefined;
};

const findLinkPrimaryAttribute = (link: ITypedTagProps<TagName.link>) => {
  const foundAttr = primaryLinkAttributes.find(attr => {
    return link.tagProps[attr] !== undefined ? !(attr === "rel" && link.tagProps[attr] === "stylesheet") : false;
  });

  if (foundAttr !== undefined) {
    return `${foundAttr}_${link.tagProps[foundAttr]}`;
  }

  return undefined;
};

const mergeAllToOne = <T extends keyof HelmetTags, TElement extends ArrayElement<HelmetTags[T]>>(
  values: TElement[] | undefined,
  isEmptyState: boolean,
  result: TElement | undefined,
  usePrevResult?: boolean,
  emptyStateFallback?: (result: TElement | undefined) => TElement | undefined
): TElement | undefined => {
  if (isEmptyState) {
    if (emptyStateFallback !== undefined) {
      return emptyStateFallback(result);
    }

    return undefined;
  }

  return (values || []).reduce((prev, current) => {
    return usePrevResult ? {...prev, ...current} : {...current};
  }, result);
};

const mergeAllToAll = <T extends keyof HelmetTags, TElement extends ArrayElement<HelmetTags[T]>>(
  values: TElement[] | undefined,
  isEmptyState: boolean,
  result: TElement[]
): TElement[] => {
  if (isEmptyState) {
    _.clear(result);
  }

  if (values) {
    result.push(...values);
  }

  return result;
};

const mergeAllByPrimaryAttribute = <T extends keyof HelmetTags, TElement = HelmetTags[T]>(
  values: TElement[] | undefined,
  isEmptyState: boolean,
  result: TElement[],
  primaryAttributeSelector: (tag: TElement) => string | undefined
) => {
  if (isEmptyState) {
    _.clear(result);
  } else {
    if (values) {
      if (result.length === 0) {
        result.push(...values);
      } else {
        const instanceGrouped = _.groupBy(
          values,
          primaryAttributeSelector,
          (item, index) => [item, index] as [TElement, number]
        );
        const resultGrouped = _.groupBy(
          result,
          primaryAttributeSelector,
          (item, index) => [item, index] as [TElement, number]
        );

        for (const [attr, instanceTags] of instanceGrouped.entries()) {
          const resultTags = resultGrouped.get(attr);

          if (resultTags === undefined) {
            result.push(...instanceTags.map(([tags]) => tags));
          } else if (instanceTags.length !== resultTags.length) {
            result = result.filter(m => !resultTags.some(([tag]) => m === tag));
            result.push(...instanceTags.map(([tag]) => tag));
          } else {
            for (let i = 0; i < instanceTags.length; i++) {
              const [instanceTag] = instanceTags[i];
              const [resultTag, resultIndex] = resultTags[i];

              result[resultIndex] = {...resultTag, ...instanceTag};
            }
          }
        }
      }
    }
  }

  return result;
};

export const buildState = (instances: IHelmetInstanceState[], priority: TagPriorityConfig[] | boolean): IHelmetState => {
  const state: IHelmetState = {
    titleTag: undefined,
    baseTag: undefined,
    bodyAttributes: undefined,
    htmlAttributes: undefined,
    metaTags: [],
    styleTags: [],
    scriptTags: [],
    linkTags: [],
    noscriptTags: [],
    headerTags: []
  };

  const titleEmptyStateFallback = (result?: ITypedTagProps<TagName.title>): ITypedTagProps<TagName.title> | undefined => {
    return result !== undefined ? {tagType: TagName.title, tagProps: {children: result.tagProps.children}} : undefined
  };

  for (const instance of instances) {
    const isEmptyState = _.isEmptyArray(instance.title, instance.link, instance.base, instance.body,
      instance.html, instance.style, instance.script, instance.noscript, instance.meta, instance.link);

    state.titleTag = mergeAllToOne(instance.title, isEmptyState, state.titleTag, true, titleEmptyStateFallback);
    state.baseTag = mergeAllToOne(instance.base, isEmptyState, state.baseTag);
    state.bodyAttributes = mergeAllToOne(instance.body, isEmptyState, state.bodyAttributes);
    state.htmlAttributes = mergeAllToOne(instance.html, isEmptyState, state.htmlAttributes);
    state.styleTags = mergeAllToAll(instance.style, isEmptyState, state.styleTags);
    state.scriptTags = mergeAllToAll(instance.script, isEmptyState, state.scriptTags);
    state.noscriptTags = mergeAllToAll(instance.noscript, isEmptyState, state.noscriptTags);
    state.metaTags = mergeAllByPrimaryAttribute(
      instance.meta,
      isEmptyState,
      state.metaTags,
      findMetaPrimaryAttribute
    );
    state.linkTags = mergeAllByPrimaryAttribute(
      instance.link,
      isEmptyState,
      state.linkTags,
      findLinkPrimaryAttribute
    );
  }

  const priorityConfig = typeof priority == "boolean"
    ? priority ? DefaultTagPriorityConfig : undefined
    : priority;

  if (priorityConfig) {
    state.headerTags = buildTagsPriority(state, priorityConfig)
  }

  return state;
};

const buildTagsPriority = (state: IHelmetState, priorityConfig: TagPriorityConfig[]): TypedTagProps[] => {
  let priorities: TypedTagProps[] = []
  for (const config of priorityConfig) {
    switch (config.tagName) {
      case TagName.title:
        if (state.titleTag) {
          priorities.push(state.titleTag);
          state.titleTag = undefined;
        }
        break
      case TagName.base:
        if (state.baseTag) {
          priorities.push(state.baseTag)
          state.baseTag = undefined;
        }
        break
      case TagName.meta:
        priorities.push(...getMatchedTags(state.metaTags, config))
        break
      case TagName.style:
        priorities.push(...getMatchedTags(state.styleTags, config))
        break
      case TagName.link:
        priorities.push(...getMatchedTags(state.linkTags, config))
        break
      case TagName.script:
        priorities.push(...getMatchedTags(state.scriptTags, config))
        break
      case TagName.noscript:
        priorities.push(...getMatchedTags(state.noscriptTags, config))
        break
    }
  }
  return priorities
}

const getMatchedTags = <T extends TagName, >(tags: ITypedTagProps<T>[], config: TagPriorityConfig) => {
  const priorities: TypedTagProps[] = [];
  for (const tag of tags) {
    if (isTagMatch(tag, config)) {
      priorities.push(tag);
      _.remove(tags, tag);
    }
  }

  return priorities
}

const isTagMatch = <T, >(tag: T, config: TagPriorityConfig) => {
  for (const [key, configValue] of Object.entries(config)) {
    if (key == "tagName") {
      continue;
    }

    const tagValue = tag[key as keyof T];
    if (tagValue === undefined) {
      return false;
    }

    if (configValue["seoAnyValue"] !== undefined && configValue["seoAnyValue"]) {
      continue;
    }

    if (configValue["seoValues"]) {
      if (!configValue["seoValues"].find((m: any) => m === tagValue)) {
        return false;
      }
    }

    if (tagValue !== configValue) {
      return false;
    }
  }

  return true;
}

const createReactComponent = <T extends TagName, >(tag: ITypedTagProps<T> | undefined, fallbackTag?: ITypedTagProps<T>): JSX.Element => tag
  ? createComponent(tag)
  : (fallbackTag ? createComponent(fallbackTag) : <></>);

const createReactComponents = <T extends TagName, >(tags: ITypedTagProps<T>[]): JSX.Element[] => tags.map((m, i) => createComponent(m, i));

const fallBackTitle: ITypedTagProps<TagName.title> = {tagType: TagName.title, tagProps: {}}
export const buildServerState = (state: IHelmetState): IHelmetServerState => {
  const titleComponentCallback = () => createReactComponent(state.titleTag, fallBackTitle);
  const baseComponentCallback = () => createReactComponent(state.baseTag);
  const bodyComponentCallback = () => createReactComponent(state.bodyAttributes);
  const htmlComponentCallback = () => createReactComponent(state.htmlAttributes);
  const metaComponentsCallback = () => createReactComponents(state.metaTags);
  const styleComponentsCallback = () => createReactComponents(state.styleTags);
  const scriptComponentsCallback = () => createReactComponents(state.scriptTags);
  const linkComponentsCallback = () => createReactComponents(state.linkTags);
  const noscriptComponentsCallback = () => createReactComponents(state.noscriptTags);

  const priorityComponentsCallback = () => createReactComponents(state.headerTags);

  return {
    title: {
      toComponent: titleComponentCallback,
      toString: () => renderToStaticMarkup(titleComponentCallback())
    },
    base: {
      toComponent: baseComponentCallback,
      toString: () => renderToStaticMarkup(baseComponentCallback())
    },
    bodyAttributes: {
      toComponent: () => bodyComponentCallback().props ?? {},
      toString: () => renderToStaticMarkup(bodyComponentCallback()).replace("<body ", "").replace("></body>", "")
    },
    htmlAttributes: {
      toComponent: () => htmlComponentCallback().props ?? {},
      toString: () => renderToStaticMarkup(htmlComponentCallback()).replace("<html ", "").replace("></html>", "")
    },
    meta: {
      toComponent: metaComponentsCallback,
      toString: () => renderToStaticMarkup(<>{metaComponentsCallback()}</>)
    },
    style: {
      toComponent: styleComponentsCallback,
      toString: () => renderToStaticMarkup(<>{styleComponentsCallback()}</>)
    },
    script: {
      toComponent: scriptComponentsCallback,
      toString: () => renderToStaticMarkup(<>{scriptComponentsCallback()}</>)
    },
    link: {
      toComponent: linkComponentsCallback,
      toString: () => renderToStaticMarkup(<>{linkComponentsCallback()}</>)
    },
    noscript: {
      toComponent: noscriptComponentsCallback,
      toString: () => renderToStaticMarkup(<>{noscriptComponentsCallback()}</>)
    },
    priority: {
      toComponent: priorityComponentsCallback,
      toString: () => renderToStaticMarkup(<>{priorityComponentsCallback()}</>)
    }
  }
}

const renderToHtmlMarkup = (node: ReactElement) => {
  return renderToStaticMarkup(node);
}

export const renderToHtmlElement = (node: ReactElement, selector: TagNames) => {
  const parser = new DOMParser();
  const htmlMarkup = renderToHtmlMarkup(node);
  const parsed = parser.parseFromString(htmlMarkup, "application/xml");
  const element = parsed.querySelector(selector);
  if (element == null) {
    throw new Error(`Couldn't found element ${selector} in the ${htmlMarkup}`)
  }

  return element;
}

export const createComponent = <T extends TagName, >(tag: ITypedTagProps<T>, key?: Key): JSX.Element => {
  return createElement(tag.tagType, {...tag.tagProps, "data-rh": true, key: key})
}

export const getHtmlAttributesFromHtmlElement = <T extends Element>(htmlElement: T) => {
  return htmlElement.getAttributeNames().reduce((result, attrName) => {
    const attrValue = htmlElement.getAttribute(attrName);
    if (attrValue !== null) {
      result.push({name: attrName, value: attrValue});
    }
    return result;
  }, [] as { name: string, value: string }[])
}



