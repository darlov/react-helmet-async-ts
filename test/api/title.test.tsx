import { customRender } from './utils';
import { Helmet, Meta, Title } from "../../src";
import { FC } from "react";

const TextComponent: FC<{ valueLeft: string, valueRight: string }> = ({
                                                                        valueLeft,
                                                                        valueRight
                                                                      }) => <>{ valueLeft } { valueRight }</>;

describe('title', () => {
  describe('API', () => {
    it('updates page title', () => {
      customRender(
        <Helmet defaultTitle="Fallback">
          <Title>Test Title</Title>
        </Helmet>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('updates page title and allows children containing expressions', () => {
      const someValue = 'Some Great Title';

      customRender(
        <Helmet>
          <Title>Title: { someValue }</Title>
        </Helmet>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('updates page title with multiple children', () => {
      customRender(
        <div>
          <Helmet>
            <Title>Test Title</Title>
          </Helmet>
          <Helmet>
            <Title>Child One Title</Title>
          </Helmet>
          <Helmet>
            <Title>Child Two Title</Title>
          </Helmet>
        </div>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('sets title based on deepest nested component', () => {
      customRender(
        <div>
          <Helmet>
            <Title>Main Title</Title>
          </Helmet>
          <Helmet>
            <Title>Nested Title</Title>
          </Helmet>
        </div>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('sets title using deepest nested component with a defined title', () => {
      customRender(
        <div>
          <Helmet>
            <Title>Main Title</Title>
          </Helmet>
          <Helmet/>
        </div>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('does not encode all characters with HTML character entity equivalents', () => {
      const chineseTitle = '膣膗 鍆錌雔';

      customRender(
        <Helmet>
          <Title>{ chineseTitle }</Title>
        </Helmet>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('uses defaultTitle if no title is defined', () => {
      customRender(
        <Helmet defaultTitle="Fallback"/>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('uses defaultTitle component if no title is defined', () => {
      customRender(
        <Helmet defaultTitle={ <TextComponent valueLeft={ "LeftFallback" } valueRight={ "RightFallback" }/> }/>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('page title with prop itemProp', () => {
      customRender(
        <Helmet defaultTitle={ "Fallback" }>
          <Title itemProp="name">Test Title with itemProp</Title>
        </Helmet>
      );

      const title = document.getElementsByTagName('title')[0];

      expect(document.title).toMatchSnapshot();
      expect(title.getAttribute('itemprop')).toBe('name');
    });

    it('retains existing title tag when no title tag is defined', () => {
      document.head.innerHTML = `<title>Existing Title</title>`;

      customRender(
        <Helmet>
          <Meta name="keywords" content="stuff"/>
        </Helmet>
      );

      expect(document.title).toMatchSnapshot();
    });

    it('clears title tag if empty title is defined', () => {
      customRender(<>
          <Helmet>
            <Title>Existing Title</Title>
            <Meta name="keywords" content="stuff"/>
          </Helmet>
          <Helmet>
            <Title />
            <Meta name="keywords" content="stuff"/>
          </Helmet>
        </>
      );
      expect(document.title).toBe('');
    });
  });
});
