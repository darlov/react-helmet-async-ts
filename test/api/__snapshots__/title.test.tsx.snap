// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`title > API > does not encode all characters with HTML character entity equivalents 1`] = `"膣膗 鍆錌雔"`;

exports[`title > API > page title with prop itemProp 1`] = `"Test Title with itemProp"`;

exports[`title > API > retains existing title tag when no title tag is defined 1`] = `"Existing Title"`;

exports[`title > API > sets title based on deepest nested component 1`] = `"Nested Title"`;

exports[`title > API > sets title using deepest nested component with a defined title 1`] = `"Main Title"`;

exports[`title > API > updates page title 1`] = `"Test Title"`;

exports[`title > API > updates page title and allows children containing expressions 1`] = `"Title: Some Great Title"`;

exports[`title > API > updates page title with multiple children 1`] = `"Child Two Title"`;

exports[`title > API > uses defaultTitle component if no title is defined 1`] = `"LeftFallback RightFallback"`;

exports[`title > API > uses defaultTitle if no title is defined 1`] = `"Fallback"`;
